---
title: "Week2: playing with annotations"
author: "Pierre-Luc"
date: "3/4/2023"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(ensembldb)
  library(GenomicRanges)
  library(epiwraps)
})
```

# 1. Fetching the gene annotation data

The following objects are downloaded below:
1. dnasm_dorso
The  drosophila melanogaster genome sequence ( dna_sm ) in TwoBit/2bit format for BDGP6.

2. ensdb_mus
The mouse (Mus Musculus) EnsDb object, version 102, genome build GRCm38.

3.dnasm_mus
The mouse genome sequence ( dna_sm ) in TwoBit/2bit format for GRCm38.
```{r}
ah <- AnnotationHub()

#Retrieve the hits for Dorsophilia melagoaster genome sequence
#from the AnnotationHub
q <- query(ah, c("Drosophila melanogaster", "ensembl", "BDGP6","2bit","dna_sm"))
q
#save the AH49674 in dnasm_dorso object
dnasm_dorso <- ah[["AH49674"]]
dnasm_dorso
```
```{r}
#Retrive the hits for Mus Musculus Ensdb object
q2 <- query(ah, c("Mus Musculus", "ensDb", "version 102", "GRCm38"))
q2
#Save the result in ensdb_mus object
ensdb_mus <- ah[["AH89211"]]
ensdb_mus
```
```{r}
q3 <- query(ah, c("Mus Musculus", "dna_sm", "2bit", "GRCm38"))
q3
dnasm_mus <- ah[["AH88477"]]

dnasm_mus
```


```{r}
# extract transcripts and genes:
txs <- transcripts(ensdb_mus)
gns<-genes(ensdb_mus)

```
# 2. Reading the ensembl object
## How many different ensembl gene IDs and gene symbols are there for protein-coding genes?


```{r, include=FALSE}
#seqlevels(txs)
#head(txs$tx_biotype)
#head(txs[,"tx_biotype"])
#table(seqnames(txs))
#shows you the available filters
#supportedFilters()
```


```{r}
table(txs$tx_biotype)


#filter with the attribute "protein_coding"
txs_pc <- transcripts(ensdb_mus, filter=TxBiotypeFilter("protein_coding"))

print(paste0("Number of protein coding genes: ",length(txs_pc)))

```
## Plot the distribution of the (spliced) length of protein-coding transcripts
First, we extract exons of protein-coding transcripts from the database. Then, we split them by transcript, before summing the width of the exons of each transcript.

```{r, include=FALSE}
#return exons
exs <- exons(ensdb_mus, column=c("tx_id"))

# same as "exons()" but returning the exons split by transcripts
exsPerTx <- exonsBy(ensdb_mus, column=c("tx_id","tx_biotype"))
```


```{r}
# we limit ourselves to protein-coding transcripts:
exsPerTx <- exonsBy(ensdb_mus, column=c("tx_id","tx_biotype"), filter=TxBiotypeFilter("protein_coding"))
```


```{r, include=FALSE}
# what are the transcripts that have more than on exon / sub-element:
head(which(lengths(exsPerTx)>1))

exsPerTx[["ENSMUST00000000003"]]
```

How many transcripts do protein-coding genes have?

```{r}
nbExonsPerPCtx <- lengths(exsPerTx)
hist(nbExonsPerPCtx, breaks=100)
```

How long are the protein coding genes?
```{r}
#retrieve all the lengths of exons for each transcript
exPerTx_widths<-width(exsPerTx)
#sum over all the rows
spliced_lengths<-sum(exPerTx_widths)
max(spliced_lengths)
#plot the histogram
hist(spliced_lengths, breaks = 100)
```


